improver/utilities/field_texture.py

```diff
@@ -0,0 +1,173 @@
+        )
+        potential_transitions = SquareNeighbourhood(sum_or_fraction="sum").run(
+            cube, radius=radius
+        )
+        potential_transitions.data = np.where(
+            cube.data > 0, potential_transitions.data * 4, 0
+        )
+
+        # Note that where there is no cloud the value is forced to 1. So high values
+        # correspond to scattered cloud or no cloud.
+
+        actual_transitions = SquareNeighbourhood(sum_or_fraction="sum").run(
+            result_cube, radius=radius
+        )
+        actual_transitions.data = np.where(cube.data > 0, actual_transitions.data, 0)
+        ratio = actual_transitions.copy(data=np.ones_like(actual_transitions.data))
+        np.divide(
+            actual_transitions.data,
+            potential_transitions.data,
+            out=ratio.data,
+            where=(potential_transitions.data > 0),
+        )
+        return ratio
+
+    @staticmethod
+    def _calculate_transitions(data):
+        padded_data = np.pad(data, 1, mode="edge")
+        diff_x = np.abs(np.diff(padded_data, axis=1))
+        diff_y = np.abs(np.diff(padded_data, axis=0))
+        cell_sum_x = diff_x[:, 0:-1] + diff_x[:, 1:]
+        cell_sum_y = diff_y[0:-1, :] + diff_y[1:, :]
+        cell_sum = cell_sum_x[1:-1, :] + cell_sum_y[:, 1:-1]
+        cell_sum = np.where(data > 0, cell_sum, 0)
+        return cell_sum
+
+    def _calculate_clumpiness(self, input_cube):
+        """
+        Calculates a field of cloud texture to use in differentiating solid and
+        more scattered cloud.
+
+        Args:
+            input_cube (cube):
+                Input data in cube format with multiple-realizations.
+
+        Returns:
+            clumpiness (cube):
+                A cube of binary data, where 1 represents sunlight and 0
+                represents cloud.
+        """
+        ratios = iris.cube.CubeList()
+
+        try:
+            cslices = input_cube.slices_over("realization")
+        except CoordinateNotFoundError:
+            cslices = [input_cube]
+        for cslice in cslices:
+            ratios.append(self._calculate_ratio(cslice, self.nbhood_radius))
+        ratios = ratios.merge_cube()
+        thresholded = BasicThreshold(self.ratio_threshold).process(ratios)
+        thresholded = iris.util.squeeze(
+            collapsed(thresholded, "realization", iris.analysis.MEAN)
+        )
+        return thresholded
+
+    def process(self, input_cube):
+        """
+        Calculates the ratio of actual to potential transitions over a binary
+        field and then produces threshold texture outputs.
+
+        Args:
+            input_cube (cube):
+                Input data in cube formate with multiple-realizations.
+
+        Returns:
+            threshold (cube)
+                Mean of the thresholded ratios in cube format.
+        """
+
+        thresholded = self._calculate_clumpiness(input_cube)
```

Comment:
I suggest you rename the calculate_clumpiness method as process and remove this process method which does nothing but call calculate_clumpiness.

Reply:
Good point. Done.